// @tkey/service-provider-torus/dist/serviceProviderTorus.esm.js
import _asyncToGenerator from '@babel/runtime/helpers/asyncToGenerator';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _get from '@babel/runtime/helpers/get';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _regeneratorRuntime from '@babel/runtime/regenerator';
import { ServiceProviderBase } from '@tkey/service-provider-base';
import CustomAuth from '@toruslabs/customauth';
import BN from 'bn.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var TorusServiceProvider = /*#__PURE__*/function (_ServiceProviderBase) {
  _inherits(TorusServiceProvider, _ServiceProviderBase);
  var _super = _createSuper(TorusServiceProvider);
  function TorusServiceProvider(_ref) {
    var _this;
    var _ref$enableLogging = _ref.enableLogging,
      enableLogging = _ref$enableLogging === void 0 ? false : _ref$enableLogging,
      postboxKey = _ref.postboxKey,
      customAuthArgs = _ref.customAuthArgs;
    _classCallCheck(this, TorusServiceProvider);
    _this = _super.call(this, {
      enableLogging: enableLogging,
      postboxKey: postboxKey
    });
    _defineProperty(_assertThisInitialized(_this), "directWeb", void 0);
    _defineProperty(_assertThisInitialized(_this), "singleLoginKey", void 0);
    _defineProperty(_assertThisInitialized(_this), "customAuthArgs", void 0);
    _this.customAuthArgs = customAuthArgs;
    _this.directWeb = new CustomAuth(customAuthArgs);
    _this.serviceProviderName = "TorusServiceProvider";
    return _this;
  }
  _createClass(TorusServiceProvider, [{
    key: "init",
    value: function () {
      var _init = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(params) {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt("return", this.directWeb.init(params));
              case 1:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
      function init(_x) {
        return _init.apply(this, arguments);
      }
      return init;
    }()
  }, {
    key: "triggerLogin",
    value: function () {
      var _triggerLogin = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(params) {
        var obj;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.directWeb.triggerLogin(params);
              case 2:
                obj = _context2.sent;
                this.postboxKey = new BN(obj.privateKey, "hex");
                return _context2.abrupt("return", obj);
              case 5:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));
      function triggerLogin(_x2) {
        return _triggerLogin.apply(this, arguments);
      }
      return triggerLogin;
    }()
  }, {
    key: "triggerAggregateLogin",
    value: function () {
      var _triggerAggregateLogin = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3(params) {
        var obj;
        return _regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.directWeb.triggerAggregateLogin(params);
              case 2:
                obj = _context3.sent;
                this.postboxKey = new BN(obj.privateKey, "hex");
                return _context3.abrupt("return", obj);
              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));
      function triggerAggregateLogin(_x3) {
        return _triggerAggregateLogin.apply(this, arguments);
      }
      return triggerAggregateLogin;
    }()
  }, {
    key: "triggerHybridAggregateLogin",
    value: function () {
      var _triggerHybridAggregateLogin = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee4(params) {
        var obj, aggregateLoginKey;
        return _regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.directWeb.triggerHybridAggregateLogin(params);
              case 2:
                obj = _context4.sent;
                aggregateLoginKey = obj.aggregateLogins[0].privateKey;
                this.postboxKey = new BN(aggregateLoginKey, "hex");
                this.singleLoginKey = new BN(obj.singleLogin.privateKey, "hex");
                return _context4.abrupt("return", obj);
              case 7:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));
      function triggerHybridAggregateLogin(_x4) {
        return _triggerHybridAggregateLogin.apply(this, arguments);
      }
      return triggerHybridAggregateLogin;
    }()
  }, {
    key: "toJSON",
    value: function toJSON() {
      return _objectSpread(_objectSpread({}, _get(_getPrototypeOf(TorusServiceProvider.prototype), "toJSON", this).call(this)), {}, {
        serviceProviderName: this.serviceProviderName,
        customAuthArgs: this.customAuthArgs
      });
    }
  }], [{
    key: "fromJSON",
    value: function fromJSON(value) {
      var enableLogging = value.enableLogging,
        postboxKey = value.postboxKey,
        customAuthArgs = value.customAuthArgs,
        serviceProviderName = value.serviceProviderName;
      if (serviceProviderName !== "TorusServiceProvider") return undefined;
      return new TorusServiceProvider({
        enableLogging: enableLogging,
        postboxKey: postboxKey,
        customAuthArgs: customAuthArgs
      });
    }
  }]);
  return TorusServiceProvider;
}(ServiceProviderBase);

export { TorusServiceProvider, TorusServiceProvider as default };
